libcperciva/datastruct/elasticarray.c

```diff
@@ -25,12 +25,15 @@ resize(struct elasticarray * EA, size_t nsize)
 	void * nbuf;
 
 	/* Figure out how large an allocation we want. */
-	if (EA->alloc < nsize) {
+	if (nsize > EA->alloc) {
 		/* We need to enlarge the buffer. */
-		nalloc = EA->alloc * 2;
+		if (EA->alloc < SIZE_MAX / 2)
+			nalloc = EA->alloc * 2;
+		else
+			nalloc = nsize;
 		if (nalloc < nsize)
 			nalloc = nsize;
-	} else if (EA->alloc > nsize * 4) {
```

Comment:
I'm much more suspicious about this code.  Unless I've made a mistake in my investigation, there's multiple cases where this produces bad values.

(all examples with `SIZE_MAX` = 9)

For example, if we have:
```
EA->alloc = 5
nsize = 5
```
Then we end up with `5 * 4 mod 10 = 0`, so we take this case, but then calculate `nalloc = nsize * 2 = 5 * 2 mod 10 = 0`, which is not at all what we want!

Maybe that's a silly example; why would we have resizing to the same size?

Consider another one:
```
EA->alloc = 7
nsize = 6
```

We're checking `nsize * 4 = 6 * 4 mod 10 = 4`, so we take this case; then set `nalloc = 6 * 2 mod 10 = 2`.

I'm going to break for lunch, then double-check this.

(I normally wouldn't post a half-finished investigation, but I thought it might be useful to see what I meant by "finding a counter-example" in my response to the other code in this function.)

Reply:
Agree. As I said in the reply to the previous comment, the 'if' block is not buggy but this 'else if' block definitely looks buggy which is why I changed this to division instead. But I definitely agree that the 'if' block's changes is not really needed as it was not buggy but the main motivation was to be more explicit as I needed to change this 'else if' block anyway.

Let me know your final thoughts. Good stuff!

Reply:
Changing it to
```
if (EA->alloc / 4 > nsize)
```
would succeed in making it safe, but division is slower than multiplication.

In this case, how likely is this going to be a problem?
- on a 64-bit system, this would mean requesting over 4.6e18 bytes (or 4.6 million TB).  That's never going to happen.
- on the other hand, `SIZE_MAX` is only required to be at least 65535.  That would mean we'd run into this problem with only 16 kB, which is an entirely reasonable size for an elastic array.
- A more plausible "old or low-powered system" would be 32 bits, which would put the "problematic request" at 1 GB.  That's unlikely, but not impossible.

Right now, I'm leaning towards keeping `(EA->alloc > nsize * 4)` and doing the `nalloc = nsize * 2` arithmetic.  But then adding

```
/* If nsize is huge, the above calculations might be wrong. */
if (nsize > SIZE_MAX / 4)
        nalloc = nsize;
```

The compiler will calculate `SIZE_MAX / 4`, so it's a simple `if (nsize > WHATEVER)` check.  And hopefully the branch prediction will figure out that `nsize` is not likely to be huge; in which case, the performance hit would be negligible.

Reply:
Good discussion. Is the division going to be problematic in practice perf wise? Although I did not think it through, if so, could we do right shift instead of division like: else if ((EA->alloc >> 2) > nsize)? What do you think?

Also it might make sense to think about the original essence of the existing algorithm. I think if we assume that there is no overflow in the 'else if' case, then the original algorithm meant if EA->alloc is equal or greater than nsize i.e., the if condition is not taken, and if it _really_ was greater than nsize * 4, then down the line a realloc would happen of size nsize * 2. But now that the else if block could be taken (according to existing code) even though EA->alloc is not _really_ greater than nsize * 4 and if we use the new if(nsize > SIZE_MAX / 4) condition as you suggested, a realloc call might happen down below. But it really meant to go to the last 'else' block which doesn't cause any realloc system call.

Let me know what you think.

Reply:
Oops, my bad on the "division" thing.  Since we're dealing with unsigned integers, I think it's safe to assume that the compiler would implement it as `EA->alloc >> 2`, so speed isn't a concern.

With that in mind, I'm actually leaning towards `EA->alloc / 4` after all.

...

I kind-of feel like an undergraduate physics student doing thought experiments about the speed of light.  "Suppose that we are in a space ship travelling at 0.9*c*, but now we want to decelerate to 0.26*c*.  Should we resize the array?"

... yeah, the closest match to the original algorithm is to do nothing.  So `EA->alloc / 4` would be the best thing.

So I now think that it's best to go with one commit which does
```
-       } else if (EA->alloc > nsize * 4) {
+       } else if (EA->alloc / 4 > nsize) {
```

and explains why in the commit message.

Reply:
Great stuff! Separate commit fixing the bug in [6b2e3cb](https://github.com/Tarsnap/tarsnap/pull/592/commits/6b2e3cbba613a23e18f794613d4444eb12beca73).